### ISSUE TITLE
OpenAPI spec

### ISSUE THREAD

**Issue by tomdavidson** [rancher/rancher#12925]
Hello, I have found https://github.com/rancher/api-spec but there does not seem to be a machine readable OpenAPI/Swagger spec for the Rancher API. I do not see an issue in 2.0 or 2.1 milestones.

Can OpenAPI be on the horizon for one of the 2.x releases?

**Comment by alexandrem** [comment]
Would be interested about this.

**Comment by cjellick** [comment]
As a stop-gap, note that if you are programming in go, there are clients here:
https://github.com/rancher/types/tree/master/client

And if you are programming in python, we have a dynamic client. You can find examples of its usage here:
https://github.com/rancher/rancher/tree/master/tests/core

**Comment by loganhz** [comment]
Please let us know in a comment if you can still reproduce the issue, and we'll reopen it

**Comment by tomdavidson** [comment]
Still do not have a known and open standard for machine readable contract. Still have not seen it on a roadmap. Hard to to take integration serious without the standardized contract.

- Here's another ask https://github.com/rancher/rancher/issues/13749
- Kqueen certainly has OpenAPI spec https://github.com/Mirantis/kqueen/blob/master/kqueen/blueprints/api/api3_0.yml

**Comment by bradjones1** [comment]
For those of us who would be using an SDK potentially auto-generated from such a spec, I'm curious if there is, at least in the short term, anything the team could share around the structure of the API, e.g. if it follows any particular pattern that we can piggyback off of in building stopgap integrations? E.g., the responses look similar in structure and content to json:api, however it doesn't conform to that standard. It looks like the api spec was forked from GoDaddy's at some point, however I don't see anything in the specification document regarding overall structure. Any info on this would be much appreciated.

I did take a stab at importing some responses from the api into swaggerhub, however given the sheer volume of models and relationships in the Rancher api, I think the only practical way to generate a spec would be using some sort of automation.

As always thanks for the great product and the help!

**Comment by bradjones1** [comment]
@vincent99 Thanks for the detail.

**Comment by ShadowJonathan** [comment]
OpenAPI/Swagger is industry standard, it's already a bit unorthodox that nowhere a dry API spec can be found, and something resembling it can only be found when installing a local rancher instance, so defining an exhaustive list of API resources and operations can help make rancher a gateway to managing clusters in a central way.

**Comment by mlebihan** [comment]
I do agree with ShadowJohathan. OpenAPI/Swagger client generator is one of the most useful one created this ten last years with _Liquibase_ and _Git_.
Tagging your server REST methods and their input/output objects with OpenAPI annotations is enough to make OpenAPI export tool generates a json describing them all, and then the OpenAPI generator tool can produce a client API able to exchange with your REST server in almost any language : I use the generator to avoid clumsy manual build requests and above all : the decoding of the response manually.

I've read here and there that some believes that REST services could continue be called with curl commands, or with a special A' in Go client that the A project built for it's project, and a B' client that the B project built in Python for it's B project, but No : OpenAPI/Swagger overrides all : it's enough by itself to ensure any exchange between any REST server written in any language and it produce a client API in any language.

OpenAPI compliance isn't an option. It's a must.

**Comment by KornKalle** [comment]
Would also love to see this.
We wanted to build some kind of automation / scripts for speeding up some repetitive tasks and just got bummed because we kind of thought there would be a standardized API documentation.

**Comment by jcamu** [comment]
Personally i take a time looking in the rancher docs, one internet , somewhere were i would be able to found API documentation.
Finally, i was surprised to not found precise documentation, and i thinks the same, the openAPI/Swagger spec/tools association works very well and permit to have a generic solution to read, interact with APIs.

**Comment by Toothwitch** [comment]
+1
We wanted to use Rancher in our Ansible Infrastructure Playbooks and looked for a proper API Spec. 
I was quite sad to see this doc design when the frontend ui is really nice and agree with many opinions here.
Every tool in our company has or has gotten an OpenAPI spec in the past years and it made working across all APIs much easier.
Explaining someone new to the company has gone down to "This is swagger, it tells you what is possible, all tools have it" with only few questions afterwards. 
I'd wish to see an OpenAPI implementation in this project to speed up the development of many of us here, instead of fighting documentation, as this is still a great product to use. Until then we will try to adapt the currently available information for our use case.

**Comment by rolandchanv** [comment]
Agreed. When we are testing how Rancher behaves when integrated in our environment, being able to use the same tools and workflows as we have for every other API would be very helpful.

We can do this for the downstream clusters and native k8s functionality, but not for the Rancher elements.

**Comment by PaulVerhoeven1** [comment]
+1

**Comment by aaronkjones** [comment]
I am looking to leverage kubeconform (successor of kubeval, sort of) to validate in CI/CD custom resources. From what I gather, if an OpenAPI spec was available, it could be converted to a JSON schema and used for this purpose. For Kubernetes, that is available [here](https://github.com/kubernetes/kubernetes/blob/master/api/openapi-spec/swagger.json) and is converted via [instrumenta/openapi2jsonschema](https://github.com/instrumenta/openapi2jsonschema) or similar for use in kubeconform.

**Comment by yyefet** [comment]
+1

**Comment by yyefet** [comment]
+1

**Comment by stefanlasiewski** [comment]
We are an enterprise customer. API documentation would help us with some of our tools that interface with Rancher & the API. We have have been hindered at the lack of documentation and mystery of certain Rancher behaviors and would really appreciate having more information for troubleshooting.

Folks who are voting/commenting here should also follow https://github.com/rancher/rancher-docs/issues/583 which seems to be the new parent issue to track API documentation.

**Comment by synec** [comment]
+1
